File: dotnet/Allmark/Render/ConsoleFootnoteRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class ConsoleFootnoteRenderer
{
    public static Renderer Create(string style)
    {
        return new Renderer
        {
            Name = "footnote",
            Render = (node, state, first, last, decode) => Render(node, state, style),
        };
    }

    public static void Render(MarkdownNode node, RendererState state, string style)
    {
        if (state.Footnotes.FirstOrDefault(f => f.Info == node.Info) == null)
        {
            state.Footnotes.Add(node);
        }
        var label = state.Footnotes.Count;
        state.Output.Append($"{style}[{label}]{RenderToConsole.AnsiReset}");
    }
}

File: swift/Sources/allmark/RenderConsole/ConsoleFootnoteRenderer.swift
import Foundation

@MainActor
let consoleFootnoteRenderer = Renderer(
	name: "footnote",
	render: renderConsoleFootnote
)

@MainActor
func renderConsoleFootnote(_ node: MarkdownNode, _ state: inout RendererState, _ first: Bool?, _ last: Bool?, _ decode: Bool?) {
	let style = ansiDim
	if state.footnotes.first(where: { $0.info == node.info }) == nil {
		state.footnotes.append(node)
	}
	let label = state.footnotes.count
	state.output += "\(style)[\(label)]\(ansiReset)"
}
